### PR TITLE
Improve error handling, tests

### LINKED ISSUE #13: gracefully handle connection failures
While using the Qubes proxy (via the SecureDrop SDK), I ran into the following exception:

```
During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/user/.local/share/virtualenvs/securedrop-proxy-3awDxAhP/lib/python3.5/site-packages/requests/adapters.py", line 445, in send
    timeout=timeout
  File "/home/user/.local/share/virtualenvs/securedrop-proxy-3awDxAhP/lib/python3.5/site-packages/urllib3/connectionpool.py", line 638, in urlopen
    _stacktrace=sys.exc_info()[2])
  File "/home/user/.local/share/virtualenvs/securedrop-proxy-3awDxAhP/lib/python3.5/site-packages/urllib3/util/retry.py", line 398, in increment
    raise MaxRetryError(_pool, url, error or ResponseError(cause))
urllib3.exceptions.MaxRetryError: HTTPConnectionPool(host='REDACTED.onion', port=80): Max retries exceeded with url: /api/v1/token (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x737017375e10>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution',))
```

We should handle this exception and return an informative response - I propose via a 500 status code and at least the explanation "connection error" in the request body. This can be used by clients to show informative feedback to the user, e.g. we can pop up something saying "Connection Error. Are you sure your workstation is connected to the internet?" 

Once this is done, we can make the corresponding change in the SDK.

### PR DESCRIPTION
**Pull request by rmol**
## Description

Try to ensure the proxy always returns a valid JSON response. Add specific handling of upstream error responses to proxy.py, and add tests and fixtures for that.

Replace the Proxy._on_done static method with an instance method.

In main.py, bail out upon receiving invalid input JSON, instead of printing an error response and then continuing to try to use it, resulting in printing another error response.

Have main.py use the callbacks on the proxy as given, instead of always overwriting them with the defaults from callbacks.py.

Rework entrypoint.py so that it should always print a JSON response, and add tests for it.

Fixes #13.

## Testing

1. Build the `securedrop-proxy` source distribution:
    - check out this branch
    - run `python setup.py sdist`
2. Build the `.deb`:
    - in a clone of `securedrop-debian-packaging`, run `PKG_PATH=/home/user/src/securedrop-proxy/dist/securedrop-proxy-0.1.6.tar.gz PKG_VERSION=0.1.6 make securedrop-proxy` (adjust the path to your working copy of this repo as necessary)
3. Install the `.deb` in `sd-svs-buster-template`:
    - `qvm-copy ~/debbuild/packaging/securedrop-proxy_0.1.6+buster_all.deb` and specify `sd-svs-buster-template` as the target VM
    - In a shell on `sd-svs-buster-template`, run:
        - `cp ~/QubesIncoming/sd-dev/securedrop-proxy_0.1.6+buster_all.deb /tmp`
        - `sudo apt install /tmp/securedrop-proxy_0.1.6+buster_all.deb`
4. Reboot.
5. Start the SecureDrop client in `sd-svs` and confirm that the proxy is communicating with the server: submissions and replies should populate with no errors.

### CODE REVIEW

**Comment by redshiftzero** [comment]
let's do the version bump in another PR (@rmol had asked me about this privately and I said let's do the release separately) since nightlies will build the latest even if we don't cut an official release

**Comment by kushaldas** [comment]
> let's do the version bump in another PR (@rmol had asked me about this privately and I said let's do the release separately) since nightlies will build the latest even if we don't cut an official release

Ah, my suggestion of version bump is for the reviewer for QA, not for the actual PR.

File: securedrop_proxy/main.py [hunk 1]
```diff
@@ -38,6 +39,6 @@ def __main__(incoming, p):
         req.body = client_req['body']
 
     p.req = req
-    p.on_save = callbacks.on_save
-    p.on_done = callbacks.on_done
+    if p.on_save is None:
```
[the remark this answers is not in the document]
Comment (rmol) [reply securedrop_proxy/main.py]: Changed this and the other occurrence I found.

File: tests/test_main.py [hunk 2]
```diff
@@ -101,9 +103,66 @@ def on_done(res):
...
+            "path_query": "/posts?userId=1",
+            "headers": { "X-Test-Header": "th" }
+        }
+
+        def on_save(fh, res, conf):
+            pass
+
+        p = proxy.Proxy(self.conf, proxy.Req(), on_save)
+        main.__main__(json.dumps(test_input), p)
+        self.assertEqual(p.req.headers, test_input["headers"])
```
[the remark this answers is not in the document]
Comment (rmol) [reply tests/test_main.py]: It was OK -- it's really just checking the constructed request -- but I gave this test its own fixture.

**Comment by rmol** [comment]
@kushaldas Thanks for the thorough review. I think I've fixed all the callback checks, and addressed the fixture mismatch for `test_input_headers`.

File: tests/test_proxy.py [hunk 3]
```diff
@@ -167,3 +171,127 @@ def test_proxy_500_misconfiguration(self):
...
+    def make_request(self, method="GET", path_query="/", headers=None):
+        req = proxy.Req()
+        req.method = method if method else "GET"
+        req.path_query = path_query if path_query else "/"
+        req.headers = headers if headers else {"Accept": "application/json"}
+        return req
+
+    def test_cannot_connect(self):
+        """
+        Test for "502 Bad Gateway" when the server can't be reached.
```
Comment (redshiftzero) [inline tests/test_proxy.py]: great tests! note to other reviewers that this provides regression coverage for #13

**Pull request merged**